   Предыдущее занятие   |         &nbsp;          |   Следующее занятие    
:----------------------:|:-----------------------:|:----------------------:
 [Урок 17](LESSON17.MD) | [Содержание](README.MD) | [Урок 19](LESSON19.MD) 

# Урок 18. Исключения

# Требования к сдаче работы

Задание выполняется в среде программирования Idea. Результат надо разместить в своем репозитории на GitHub.
Прислать в гугл-формы(размещенная на моем сайте) ссылку на репозиторий.

***Критерии оценивания***

* Оценка 5 - надо сделать все задания.
* Оценка 4 - надо сделать любые пять заданий.
* Оценка 3 - надо сделать любые четыре задания.

## Введение в тему

Исключения — это способ Java сообщить разработчику и программе,
что что-то пошло не так во время выполнения кода.

В этой теме научим вас не бояться исключений: разберём, из-за чего они происходят,
как их обрабатывать и избегать. А ещё:

* какие исключения бывают и что такое «иерархия исключений»;
* как научить ваш код правильно реагировать на ошибки;
* как предотвратить сбои в программе из-за непредвиденных проблем;
* как гарантировать, что определённые действия всё равно будут выполнены — даже если что-то пошло не так;
* как создавать собственные исключения, чтобы ещё эффективнее управлять программой и возможными ошибками.

В качестве практики вы потренируетесь отлавливать исключения.
Это подготовит вас к нештатным ситуациям, с которыми вы можете столкнуться
в профессиональной разработке.

<details>

<summary>Иерархия исключений</summary>

## Иерархия исключений

В ходе работы программ постоянно происходят сбои: в метод передано значение неподходящего типа,
произошёл выход за границы массива, закончилась память на компьютере или возникли неполадки
у провайдера при передаче данных.
Эти и другие ошибки и нештатные ситуации в программировании называют **исключениями** (англ. _exception_).

### Наследники Throwable

Исключения — это те ошибки и сбои, которые возникают, когда программа уже запустилась,
начала работать, но вдруг что-то пошло не так.
В стандартной библиотеке Java описано множество исключений.
Кроме того, разработчики могут создавать их сами.

Все исключения наследуют от одного класса — `Throwable` (англ. «выбрасываемый»).
Только экземпляры `Throwable` или его подклассов могут быть использованы
для генерации исключений. Кроме того, класс `Throwable` содержит все основные методы
для получения информации о произошедшей ошибке.

От `Throwable` наследуют две больших группы классов.
Вершина первой — подкласс `Exception`(англ. «исключение»), а второй `Error` (англ. «ошибка»).

![img.png](img.png)

Группа `Exception` охватывает основную часть исключений. В неё входят ошибки при работе
с файлами и сетью, сбои взаимодействия с базами данных и другие.
Кроме того, от `Exception` идёт ветвь исключений с особым поведением — `RuntimeException`
(от англ. _runtime_ — «выполнение, исполнение»). К ним относятся выход за пределы массива
или неверно переданные данные в метод, ошибки при арифметических операциях
(например, деление на ноль) и обращения к неинициализированным объектам.

Группа `Error` — это классы, описывающие критические ошибки, которые в большинстве случаев
создают аварийную ситуацию и требуют перезапуска программы. Именно в результате исключений
`Error` пользователь сталкивается с сообщением типа «Извините, произошла непредвиденная ошибка!!!».
Это могут быть сбои из-за нехватки памяти —
`OutOfMemoryError` (от англ. _out of memory_ — «за пределами памяти») или проблемы
с JVM — `VirtualMachineError`.

### Проверяемые и непроверяемые

Все исключения делятся на два типа: **проверяемые** (англ. _checked_) и **непроверяемые** (англ. _unchecked_).

**Проверяемые исключения** — это исключения, обработка которых обязательна и
является частью логики приложения. К примеру, программист может предусмотреть,
что вместо числа будет передана строка, и настроить повторный запрос данных для такого случая.
Или же пользователь введёт некорректный адрес интернет-ресурса и система не может его распознать.

> К проверяемым относятся исключения, наследуемые напрямую от `Throwable`,
> а также исключения класса `Exception` и его наследников,
> кроме ветви `RuntimeException`.
> Исключения `RuntimeException` ведут себя по-другому и относятся к непроверяемым.

Таким образом, проверяемые исключения — это те, **которые ожидаемо могут случиться**:
программист эту вероятность должен учесть и написать код, который их обрабатывает.
Если не настроить обработку проверяемых исключений, то программа не скомпилируется.

```java
public class Practicum {
    public static void main(String[] args) {
        parseUrl("httpasdas");
    }

    public static java.net.URL parseUrl(final String input) {

        /* Код не скомпилируется, потому что конструктор класса генерирует 
        проверяемое исключение MalformedURLException. */

        return new java.net.URL(input);
    }
}
```

![img_1.png](img_1.png)

Сообщение об ошибке будет примерно таким — `java: unreported exception
java.net.MalformedURLException; must be caught or declared to be thrown.`
Компилятор сообщает, что исключение `MalformedURLException`
должно быть «отловлено или объявлено выброшенным».

> К непроверяемым исключениям относятся те ошибки и нештатные ситуации,
> которые не были предусмотрены логикой
> программы и их обработка необязательна. Это исключения `Error` и ветвь `RuntimeException`.

В большинстве случаев при непроверяемых исключениях работа программы немедленно прекращается.
Например, разработчик так написал код, что программа начала бесконечно вызывать какой-то метод и переполнила стек
вызовов. Или забыл
создать объект, но попытался вывести его состояние на экран — в итоге ошибка заблокирует
работу приложения.

Поскольку разработчики заранее не обрабатывают непроверяемые исключения, если программа их выбросит, то пользователь
может увидеть какое-то странное сообщение.
Код с непроверяемым исключением успешно скомпилируется, но при выполнении упадёт с ошибкой:

```java
public class Practicum {
    public static void main(String[] args) {
        print(null);
    }

    private static void print(final Object object) {
        // Код скомпилируется, но непроверяемое исключение NullPointerException нарушит его работу.
        System.out.println(object.toString());
    }
}
```

К слову, `NullPointerException` — это, наверное, наиболее часто встречающееся исключение.
`NullPointerException` вездесуще и обязательно настигнет вас.

Определить, к какому из типов относится то или иное исключение, можно либо по иерархии классов,
либо по их поведению в коде. Главное запомнить, что проверяемые обязательны к обработке,
а непроверяемые
можно как обрабатывать за счёт добавления новой логики,
так и нет — решать проблемы по мере поступления.

### Самые распространённые исключения в Java

Типов исключений очень много.

![img_2.png](img_2.png)

#### Непроверяемые исключения, которые наследуют от Error.

| Название            | Описание                                                                                                                          |
|---------------------|-----------------------------------------------------------------------------------------------------------------------------------|
| VirtualMachineError | Базовые ошибки JVM, связанные с тем, что исчерпаны ресурсы или обнаружены повреждения.                                            |
| OutOfMemoryError	   | Производный класс от VirtualMachineError, который показывает ошибки из-за нехватки памяти.                                        |
| StackOverflowError  | Производный класс от VirtualMachineError, показывает переполнение стека из-за того, что метод слишком много раз вызывал сам себя. |
| AssertionError      | Ошибка утверждения.                                                                                                               |
| IOError             | Исключение, которое происходит при серьёзных ошибках ввода-вывода.                                                                |
| ThreadDeath         | Возникает при вызове метода Thread.stop() у потока.                                                                               |

#### Проверяемые исключения, которые наследуют от Exception.

| Название              | Описание                                                                                                                                         |
|-----------------------|--------------------------------------------------------------------------------------------------------------------------------------------------|
| IOException           | Это базовый класс проверяемых исключений ввода-вывода.                                                                                           | 
| EOFException          | Исключение, которое сигнализирует о внезапном достижении конца файла или потока.                                                                 | 
| FileNotFoundException | Файл по указанному пути не существует.                                                                                                           | 
| FileSystemException   | Базовый класс для исключений файловой системы, таких как ошибка доступа, попытка создать уже существующий файл или удалить не пустую директорию. | 
| MalformedURLException | Неверный синтаксис при создании класса ссылки из строки.                                                                                         | 
| SocketException       | Ошибки при создании/обрыве соединения по сокету.                                                                                                 | 
| UnknownHostException	 | Невозможность определить IP-адрес узла по доменному имени.                                                                                       | 
| SQLException          | Ошибки при работе с базой данных.                                                                                                                | 
| TimeoutException	     | Исключение — происходит у методов, у которых ограничено выполнение по времени.                                                                   | 
| URISyntaxException    | Неверный формат универсального идентификатора ресурсов.                                                                                          | 

#### Непроверяемые исключения, которые унаследованы от RuntimeException.

| Название                      | Описание                                                                                    |
|-------------------------------|---------------------------------------------------------------------------------------------|
| ArithmeticException           | Исключения при арифметических операциях, например деление на 0.                             |
| IllegalArgumentException      | Возникает при неверно переданных в метод или конструктор параметрах.                        |
| IndexOutOfBoundsException     | Исключение при выходе за заданный диапазон.                                                 |
| NoSuchElementException        | Данного элемента больше не существует в перечислении.                                       |
| NullPointerException          | Приложение пытается использовать null в том месте, где требуется инициализированный объект. |
| UnsupportedOperationException | Операция не поддерживается вызываемым объектом.                                             |
| ClassCastException            | Невозможность привести объект к заданному типу.                                             |


</details>

   Предыдущее занятие   |         &nbsp;          |   Следующее занятие    
:----------------------:|:-----------------------:|:----------------------:
 [Урок 17](LESSON17.MD) | [Содержание](README.MD) | [Урок 19](LESSON19.MD) 